release/src/router/gettext/gettext-tools/man/msggrep.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.24.
.TH MSGGREP "1" "May 2005" "GNU gettext-tools 0.14.5" GNU
.SH NAME
msggrep \- pattern matching on message catalog
.SH SYNOPSIS
.B msggrep
[\fIOPTION\fR] [\fIINPUTFILE\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Extracts all messages of a translation catalog that match a given pattern
or belong to some given source files.
.PP
Mandatory arguments to long options are mandatory for short options too.
.SS "Input file location:"
.TP
INPUTFILE
input PO file
.TP
\fB\-D\fR, \fB\-\-directory\fR=\fIDIRECTORY\fR
add DIRECTORY to list for input files search
.PP
If no input file is given or if it is -, standard input is read.
.SS "Output file location:"
.TP
\fB\-o\fR, \fB\-\-output\-file\fR=\fIFILE\fR
write output to specified file
.PP
The results are written to standard output if no output file is specified
or if it is -.
.SS "Message selection:"
.IP
[-N SOURCEFILE]... [-M DOMAINNAME]...
[-K MSGID-PATTERN] [-T MSGSTR-PATTERN] [-C COMMENT-PATTERN]
.PP
A message is selected if it comes from one of the specified source files,
or if it comes from one of the specified domains,
or if \fB\-K\fR is given and its key (msgid or msgid_plural) matches MSGID-PATTERN,
or if \fB\-T\fR is given and its translation (msgstr) matches MSGSTR-PATTERN,
or if \fB\-C\fR is given and the translator's comment matches COMMENT-PATTERN.
.PP
When more than one selection criterion is specified, the set of selected
messages is the union of the selected messages of each criterion.
.SS "MSGID-PATTERN or MSGSTR-PATTERN or COMMENT-PATTERN syntax:"
.IP
[-E | \fB\-F]\fR [-e PATTERN | \fB\-f\fR FILE]...
.PP
PATTERNs are basic regular expressions by default, or extended regular
expressions if \fB\-E\fR is given, or fixed strings if \fB\-F\fR is given.
.TP
\fB\-N\fR, \fB\-\-location\fR=\fISOURCEFILE\fR
select messages extracted from SOURCEFILE
.TP
\fB\-M\fR, \fB\-\-domain\fR=\fIDOMAINNAME\fR
select messages belonging to domain DOMAINNAME
.TP
\fB\-K\fR, \fB\-\-msgid\fR
start of patterns for the msgid
.TP
\fB\-T\fR, \fB\-\-msgstr\fR
start of patterns for the msgstr
.TP
\fB\-C\fR, \fB\-\-comment\fR
start of patterns for the translator's comment
.TP
\fB\-E\fR, \fB\-\-extended\-regexp\fR
PATTERN is an extended regular expression
.TP
\fB\-F\fR, \fB\-\-fixed\-strings\fR
PATTERN is a set of newline-separated strings
.TP
\fB\-e\fR, \fB\-\-regexp\fR=\fIPATTERN\fR
use PATTERN as a regular expression
.TP
\fB\-f\fR, \fB\-\-file\fR=\fIFILE\fR
obtain PATTERN from FILE
.TP
\fB\-i\fR, \fB\-\-ignore\-case\fR
ignore case distinctions
.SS "Input file syntax:"
.TP
\fB\-P\fR, \fB\-\-properties\-input\fR
input file is in Java .properties syntax
.TP
\fB\-\-stringtable\-input\fR
input file is in NeXTstep/GNUstep .strings syntax
.SS "Output details:"
.TP
\fB\-\-no\-escape\fR
do not use C escapes in output (default)
.TP
\fB\-\-escape\fR
use C escapes in output, no extended chars
.TP
\fB\-\-force\-po\fR
write PO file even if empty
.TP
\fB\-\-indent\fR
indented output style
.TP
\fB\-\-no\-location\fR
suppress '#: filename:line' lines
.TP
\fB\-\-add\-location\fR
preserve '#: filename:line' lines (default)
.TP
\fB\-\-strict\fR
strict Uniforum output style
.TP
\fB\-p\fR, \fB\-\-properties\-output\fR
write out a Java .properties file
.TP
\fB\-\-stringtable\-output\fR
write out a NeXTstep/GNUstep .strings file
.TP
\fB\-w\fR, \fB\-\-width\fR=\fINUMBER\fR
set output page width
.TP
\fB\-\-no\-wrap\fR
do not break long message lines, longer than
the output page width, into several lines
.TP
\fB\-\-sort\-output\fR
generate sorted output
.TP
\fB\-\-sort\-by\-file\fR
sort output by file location
.SS "Informative output:"
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
output version information and exit
.SH AUTHOR
Written by Bruno Haible.
.SH "REPORTING BUGS"
Report bugs to <bug-gnu-gettext@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2001-2005 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
The full documentation for
.B msggrep
is maintained as a Texinfo manual.  If the
.B info
and
.B msggrep
programs are properly installed at your site, the command
.IP
.B info msggrep
.PP
should give you access to the complete manual.


CHANGELOG
Tomato USB Mod (ND version)
===========================

build 01 - 12/17/2008
---------------------

Based on the official Tomato 1.23 ND.

Added support for USB 1.1 (OHCI and UHCI) and USB 2.0.

USB support is configurable via Tomato GUI.

USB storage (Ext2, Ext3 and FAT file systems) and USB printing support.

Non-spooling printer server (p910nd 0.92) is included and started
automatically if you enable printing support. Bidirectional copying can
be disabled via GUI if it causes problems with your printer.

Fixed slow running clock problem on Asus WL-520GU. After flashing this
firmware you have to reboot the router at least once for clock to get
fixed.

UPnP is disabled by default.

Added CGI support to HTTP daemon - you can place your cgi scripts into
"/www/ext/cgi-bin" folder, and they will be executed when you access them
via http://<router_ip>/ext/cgi-bin/my_script_name.


build 02 - 12/20/2008
---------------------

Allowed changing USB settings without rebooting a router.

Fixed non-working AIR (WLAN) LED on Asus WL-520GU.

USB support code clean-up.


build 03 - 12/22/2008
---------------------

Cosmetic changes only.


build 04 - 12/26/2008
---------------------

Added FTP server daemon (vsftpd 2.0.7) configurable via GUI.

Updated kernel printer driver to version 0.13 and included some fixes
to the printer driver made by Asus.


build 05 - 12/31/2008 and 07 - 01/03/2009 (UNSTABLE)
------------------------------------------------------

Attempt to update Broadcom Wireless and Ethernet Controller driver to the
version 4.150.10.X (original Tomato included driver is version 4.130.19.0).

Added new setting to "QoS -> Basic Settings" page to select Qdisc
Scheduler (sfq | pfifo).

Auto-mounting for USB storage drives. When this feature is activated, and
USB drive is plugged in, all mountable partitions will be automatically
mounted to a new directories under /mnt. If you want to override default
mount points, mount your partition manually in the satrtup script, i.e.:
"mount -o noatime /dev/discs/disc0/part1 /opt".
If you mount your partition manually, it will not be re-mounted to a
default mount point.

This build, as well as builds 06 and 07 had several issues: using Wireless
Survey and/or changing some settings could crash the router or cause it
to hang.


build 08 - 01/07/2009
---------------------

Since both previous versions with updated WL driver were unstable, rolled
back Broadcom Wireless Controller driver to the old version 4.130.19.0
(original Tomato ND version).

Fixed several bugs introduced in the previous releases.


build 09 - 01/09/2009
---------------------

Added Network File Sharing - Samba 2.0.10.

Added USB Hotplug script - useful for configuring attached USB devices,
i.e. loading firmware into HP1018/1020 or similar printer.

Updated USB drivers to the newer version from kernel 2.4.35/37, applied
some patches to stabilize USB.

Applied patch to allow using routers with up to 128 MB RAM.


build 11 - 01/16/2009
---------------------

Upgraded Broadcom Wireless driver to version 4.150.10.5.

Included SpeedMod by Rodney Chua.

Added functionality to the GUI to unmount (safely remove) USB drives,
reworked the display of attached usb devices.

Fixed "Kernel BUG" error when transfering files to shared USB drives
using Samba.

Added "Regulatory Mode" setting (off|g|h) to "Advanced -> Wireless"
page. This solves long-standing Tomato connectivity problems for Intel
Centrino (2200bg etc) wireless cards.

Allowed up to 100 applications in UPnP.


build 12 - 01/25/2009
---------------------

Bug fixes: cosmetic fixes for GUI, using "Measure Noise Floor" function
on the Device List page should no longer disable wireless.

Improved support for Ext2/Ext3 filesystems (allow expanded inodes,
forward compatibility).

Replaced UPnP daemon with MiniUPnPd.

Restored the gpio utility temporarily removed from previous build.


build 14 - 01/27/2009
---------------------

Minor improvements for FAT filesystem support.

Improved SpeedMod implementation.


build 15 - 01/29/2009
---------------------

Bug fix: corrected firewall rules when WAN access is enabled for FTP.


build 16 - 01/31/2009
---------------------

Fixed error connecting to the FTP server when directory listings are
disabled.

Added support for mounting loopback devices. The loop device module loop.o
is not included but available separately, and can be loaded with insmod
command.

Added option to automatically share all mounted USB partitions.


build 17 - 02/04/2009
---------------------

Patch to resolve read errors on spin-up with some auto spin-down USB
drives (not tested).

Miniupnpd daemon updated to the latest version 20090129. Fixed UPnP
presentation URL when non-standard http port is used.

Fixed LEDs/buttons/GPIO bug introduced with WL driver update (only
affected some router models).

Code clean up, cosmetics.


build 18 - 02/05/2009
---------------------

Removed excessive stripping of shared libraries (fixes problem introduced
in build 17 when using IPP2P in QoS rules).

Added limited kernel-level support for national codepages needed by Samba.


build 19 - 02/10/2009
---------------------

Restored original Tomato setting to show/hide UPnP server in "My Network
Places".

Added scsi-stop command to spin down some USB drives (find a description
at http://wl500g.info/showpost.php?p=47135&postcount=10).


release/src/router/gettext/gettext-tools/src/msggrep.c
/* Extract some translations of a translation catalog.
   Copyright (C) 2001-2005 Free Software Foundation, Inc.
   Written by Bruno Haible <haible@clisp.cons.org>, 2001.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  */


#ifdef HAVE_CONFIG_H
# include "config.h"
#endif
#include <alloca.h>

#include <assert.h>
#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#elif defined _MSC_VER || defined __MINGW32__
# include <io.h>
#endif

#include <fnmatch.h>

#include "closeout.h"
#include "dir-list.h"
#include "error.h"
#include "error-progname.h"
#include "progname.h"
#include "relocatable.h"
#include "basename.h"
#include "message.h"
#include "read-po.h"
#include "write-po.h"
#include "str-list.h"
#include "msgl-charset.h"
#include "xalloc.h"
#include "xallocsa.h"
#include "exit.h"
#include "libgrep.h"
#include "gettext.h"

#define _(str) gettext (str)


/* Force output of PO file even if empty.  */
static int force_po;

/* Selected source files.  */
static string_list_ty *location_files;

/* Selected domain names.  */
static string_list_ty *domain_names;

/* Task for each grep pass.  */
struct grep_task {
  matcher_t *matcher;
  size_t pattern_count;
  char *patterns;
  size_t patterns_size;
  bool case_insensitive;
  void *compiled_patterns;
};
static struct grep_task grep_task[3];

/* Long options.  */
static const struct option long_options[] =
{
  { "add-location", no_argument, &line_comment, 1 },
  { "comment", no_argument, NULL, 'C' },
  { "directory", required_argument, NULL, 'D' },
  { "domain", required_argument, NULL, 'M' },
  { "escape", no_argument, NULL, CHAR_MAX + 1 },
  { "extended-regexp", no_argument, NULL, 'E' },
  { "file", required_argument, NULL, 'f' },
  { "fixed-strings", no_argument, NULL, 'F' },
  { "force-po", no_argument, &force_po, 1 },
  { "help", no_argument, NULL, 'h' },
  { "ignore-case", no_argument, NULL, 'i' },
  { "indent", no_argument, NULL, CHAR_MAX + 2 },
  { "location", required_argument, NULL, 'N' },
  { "msgid", no_argument, NULL, 'K' },
  { "msgstr", no_argument, NULL, 'T' },
  { "no-escape", no_argument, NULL, CHAR_MAX + 3 },
  { "no-location", no_argument, &line_comment, 0 },
  { "no-wrap", no_argument, NULL, CHAR_MAX + 6 },
  { "output-file", required_argument, NULL, 'o' },
  { "properties-input", no_argument, NULL, 'P' },
  { "properties-output", no_argument, NULL, 'p' },
  { "regexp", required_argument, NULL, 'e' },
  { "sort-by-file", no_argument, NULL, CHAR_MAX + 4 },
  { "sort-output", no_argument, NULL, CHAR_MAX + 5 },
  { "strict", no_argument, NULL, 'S' },
  { "stringtable-input", no_argument, NULL, CHAR_MAX + 7 },
  { "stringtable-output", no_argument, NULL, CHAR_MAX + 8 },
  { "version", no_argument, NULL, 'V' },
  { "width", required_argument, NULL, 'w' },
  { NULL, 0, NULL, 0 }
};


/* Forward declaration of local functions.  */
static void no_pass (int opt)
#if defined __GNUC__ && ((__GNUC__ == 2 && __GNUC_MINOR__ >= 5) || __GNUC__ > 2)
	__attribute__ ((noreturn))
#endif
;
static void usage (int status)
#if defined __GNUC__ && ((__GNUC__ == 2 && __GNUC_MINOR__ >= 5) || __GNUC__ > 2)
	__attribute__ ((noreturn))
#endif
;
static msgdomain_list_ty *process_msgdomain_list (msgdomain_list_ty *mdlp);

#if defined(__UCLIBC__) && __UCLIBC_MINOR__ == 9 && __UCLIBC_SUBLEVEL__ == 29
void (*error_print_progname) (void) = NULL;
#endif

int
main (int argc, char **argv)
{
  int opt;
  bool do_help;
  bool do_version;
  char *output_file;
  const char *input_file;
  int grep_pass;
  msgdomain_list_ty *result;
  bool sort_by_filepos = false;
  bool sort_by_msgid = false;
  size_t i;

  /* Set program name for messages.  */
  set_program_name (argv[0]);
  error_print_progname = maybe_print_progname;

#ifdef HAVE_SETLOCALE
  /* Set locale via LC_ALL.  */
  setlocale (LC_ALL, "");
#endif

  /* Set the text message domain.  */
  bindtextdomain (PACKAGE, relocate (LOCALEDIR));
  textdomain (PACKAGE);

  /* Ensure that write errors on stdout are detected.  */
  atexit (close_stdout);

  /* Set default values for variables.  */
  do_help = false;
  do_version = false;
  output_file = NULL;
  input_file = NULL;
  grep_pass = -1;
  location_files = string_list_alloc ();
  domain_names = string_list_alloc ();

  for (i = 0; i < 3; i++)
    {
      struct grep_task *gt = &grep_task[i];

      gt->matcher = &matcher_grep;
      gt->pattern_count = 0;
      gt->patterns = NULL;
      gt->patterns_size = 0;
      gt->case_insensitive = false;
    }

  while ((opt = getopt_long (argc, argv, "CD:e:Ef:FhiKM:N:o:pPTVw:",
			     long_options, NULL))
	 != EOF)
    switch (opt)
      {
      case '\0':		/* Long option.  */
	break;

      case 'C':
	grep_pass = 2;
	break;

      case 'D':
	dir_list_append (optarg);
	break;

      case 'e':
	if (grep_pass < 0)
	  no_pass (opt);
	{
	  struct grep_task *gt = &grep_task[grep_pass];
	  /* Append optarg and a newline to gt->patterns.  */
	  size_t len = strlen (optarg);
	  gt->patterns =
	    (char *) xrealloc (gt->patterns, gt->patterns_size + len + 1);
	  memcpy (gt->patterns + gt->patterns_size, optarg, len);
	  gt->patterns_size += len;
	  *(gt->patterns + gt->patterns_size) = '\n';
	  gt->patterns_size += 1;
	  gt->pattern_count++;
	}
	break;

      case 'E':
	if (grep_pass < 0)
	  no_pass (opt);
	grep_task[grep_pass].matcher = &matcher_egrep;
	break;

      case 'f':
	if (grep_pass < 0)
	  no_pass (opt);
	{
	  struct grep_task *gt = &grep_task[grep_pass];
	  /* Append the contents of the specified file to gt->patterns.  */
	  FILE *fp = fopen (optarg, "r");

	  if (fp == NULL)
	    error (EXIT_FAILURE, errno, _("\
error while opening \"%s\" for reading"), optarg);

	  while (!feof (fp))
	    {
	      char buf[4096];
	      size_t count = fread (buf, 1, sizeof buf, fp);

	      if (count == 0)
		{
		  if (ferror (fp))
		    error (EXIT_FAILURE, errno, _("\
error while reading \"%s\""), optarg);
		  /* EOF reached.  */
		  break;
		}

	      gt->patterns =
		(char *) xrealloc (gt->patterns, gt->patterns_size + count);
	      memcpy (gt->patterns + gt->patterns_size, buf, count);
	      gt->patterns_size += count;
	    }

	  /* Append a final newline if file ended in a non-newline.  */
	  if (gt->patterns_size > 0
	      && *(gt->patterns + gt->patterns_size - 1) != '\n')
	    {
	      gt->patterns =
		(char *) xrealloc (gt->patterns, gt->patterns_size + 1);
	      *(gt->patterns + gt->patterns_size) = '\n';
	      gt->patterns_size += 1;
	    }

	  fclose (fp);
	  gt->pattern_count++;
	}
	break;

      case 'F':
	if (grep_pass < 0)
	  no_pass (opt);
	grep_task[grep_pass].matcher = &matcher_fgrep;
	break;

      case 'h':
	do_help = true;
	break;

      case 'i':
	if (grep_pass < 0)
	  no_pass (opt);
	grep_task[grep_pass].case_insensitive = true;
	break;

      case 'K':
	grep_pass = 0;
	break;

      case 'M':
	string_list_append (domain_names, optarg);
	break;

      case 'N':
	string_list_append (location_files, optarg);
	break;

      case 'o':
	output_file = optarg;
	break;

      case 'p':
	message_print_syntax_properties ();
	break;

      case 'P':
	input_syntax = syntax_properties;
	break;

      case 'S':
	message_print_style_uniforum ();
	break;

      case 'T':
	grep_pass = 1;
	break;

      case 'V':
	do_version = true;
	break;

      case 'w':
	{
	  int value;
	  char *endp;
	  value = strtol (optarg, &endp, 10);
	  if (endp != optarg)
	    message_page_width_set (value);
	}
	break;

      case CHAR_MAX + 1:
	message_print_style_escape (true);
	break;

      case CHAR_MAX + 2:
	message_print_style_indent ();
	break;

      case CHAR_MAX + 3:
	message_print_style_escape (false);
	break;

      case CHAR_MAX + 4:
	sort_by_filepos = true;
	break;

      case CHAR_MAX + 5:
	sort_by_msgid = true;
	break;

      case CHAR_MAX + 6: /* --no-wrap */
	message_page_width_ignore ();
	break;

      case CHAR_MAX + 7: /* --stringtable-input */
	input_syntax = syntax_stringtable;
	break;

      case CHAR_MAX + 8: /* --stringtable-output */
	message_print_syntax_stringtable ();
	break;

      default:
	usage (EXIT_FAILURE);
	break;
      }

  /* Version information is requested.  */
  if (do_version)
    {
      printf ("%s (GNU %s) %s\n", basename (program_name), PACKAGE, VERSION);
      /* xgettext: no-wrap */
      printf (_("Copyright (C) %s Free Software Foundation, Inc.\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\
"),
	      "2001-2005");
      printf (_("Written by %s.\n"), "Bruno Haible");
      exit (EXIT_SUCCESS);
    }

  /* Help is requested.  */
  if (do_help)
    usage (EXIT_SUCCESS);

  /* Test whether we have an .po file name as argument.  */
  if (optind == argc)
    input_file = "-";
  else if (optind + 1 == argc)
    input_file = argv[optind];
  else
    {
      error (EXIT_SUCCESS, 0, _("at most one input file allowed"));
      usage (EXIT_FAILURE);
    }

  /* Verify selected options.  */
  if (!line_comment && sort_by_filepos)
    error (EXIT_FAILURE, 0, _("%s and %s are mutually exclusive"),
	   "--no-location", "--sort-by-file");

  if (sort_by_msgid && sort_by_filepos)
    error (EXIT_FAILURE, 0, _("%s and %s are mutually exclusive"),
	   "--sort-output", "--sort-by-file");

  /* Compile the patterns.  */
  for (grep_pass = 0; grep_pass < 3; grep_pass++)
    {
      struct grep_task *gt = &grep_task[grep_pass];

      if (gt->pattern_count > 0)
	{
	  if (gt->patterns_size > 0)
	    {
	      /* Strip trailing newline.  */
	      assert (gt->patterns[gt->patterns_size - 1] == '\n');
	      gt->patterns_size--;
	    }
	  gt->compiled_patterns =
	    gt->matcher->compile (gt->patterns, gt->patterns_size,
				  gt->case_insensitive, false, false, '\n');
	}
    }

  /* Read input file.  */
  result = read_po_file (input_file);

  if (grep_task[0].pattern_count > 0
      || grep_task[1].pattern_count > 0
      || grep_task[2].pattern_count > 0)
    {
      /* Warn if the current locale is not suitable for this PO file.  */
      compare_po_locale_charsets (result);
    }

  /* Select the messages.  */
  result = process_msgdomain_list (result);

  /* Sort the results.  */
  if (sort_by_filepos)
    msgdomain_list_sort_by_filepos (result);
  else if (sort_by_msgid)
    msgdomain_list_sort_by_msgid (result);

  /* Write the merged message list out.  */
  msgdomain_list_print (result, output_file, force_po, false);

  exit (EXIT_SUCCESS);
}


static void
no_pass (int opt)
{
  error (EXIT_SUCCESS, 0,
	 _("option '%c' cannot be used before 'K' or 'T' or 'C' has been specified"),
	 opt);
  usage (EXIT_FAILURE);
}


/* Display usage information and exit.  */
static void
usage (int status)
{
  if (status != EXIT_SUCCESS)
    fprintf (stderr, _("Try `%s --help' for more information.\n"),
	     program_name);
  else
    {
      printf (_("\
Usage: %s [OPTION] [INPUTFILE]\n\
"), program_name);
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Extracts all messages of a translation catalog that match a given pattern\n\
or belong to some given source files.\n\
"));
      printf ("\n");
      printf (_("\
Mandatory arguments to long options are mandatory for short options too.\n"));
      printf ("\n");
      printf (_("\
Input file location:\n"));
      printf (_("\
  INPUTFILE                   input PO file\n"));
      printf (_("\
  -D, --directory=DIRECTORY   add DIRECTORY to list for input files search\n"));
      printf (_("\
If no input file is given or if it is -, standard input is read.\n"));
      printf ("\n");
      printf (_("\
Output file location:\n"));
      printf (_("\
  -o, --output-file=FILE      write output to specified file\n"));
      printf (_("\
The results are written to standard output if no output file is specified\n\
or if it is -.\n"));
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Message selection:\n\
  [-N SOURCEFILE]... [-M DOMAINNAME]...\n\
  [-K MSGID-PATTERN] [-T MSGSTR-PATTERN] [-C COMMENT-PATTERN]\n\
A message is selected if it comes from one of the specified source files,\n\
or if it comes from one of the specified domains,\n\
or if -K is given and its key (msgid or msgid_plural) matches MSGID-PATTERN,\n\
or if -T is given and its translation (msgstr) matches MSGSTR-PATTERN,\n\
or if -C is given and the translator's comment matches COMMENT-PATTERN.\n\
\n\
When more than one selection criterion is specified, the set of selected\n\
messages is the union of the selected messages of each criterion.\n\
\n\
MSGID-PATTERN or MSGSTR-PATTERN or COMMENT-PATTERN syntax:\n\
  [-E | -F] [-e PATTERN | -f FILE]...\n\
PATTERNs are basic regular expressions by default, or extended regular\n\
expressions if -E is given, or fixed strings if -F is given.\n\
\n\
  -N, --location=SOURCEFILE   select messages extracted from SOURCEFILE\n\
  -M, --domain=DOMAINNAME     select messages belonging to domain DOMAINNAME\n\
  -K, --msgid                 start of patterns for the msgid\n\
  -T, --msgstr                start of patterns for the msgstr\n\
  -C, --comment               start of patterns for the translator's comment\n\
  -E, --extended-regexp       PATTERN is an extended regular expression\n\
  -F, --fixed-strings         PATTERN is a set of newline-separated strings\n\
  -e, --regexp=PATTERN        use PATTERN as a regular expression\n\
  -f, --file=FILE             obtain PATTERN from FILE\n\
  -i, --ignore-case           ignore case distinctions\n\
"));
      printf ("\n");
      printf (_("\
Input file syntax:\n"));
      printf (_("\
  -P, --properties-input      input file is in Java .properties syntax\n"));
      printf (_("\
      --stringtable-input     input file is in NeXTstep/GNUstep .strings syntax\n"));
      printf ("\n");
      printf (_("\
Output details:\n"));
      printf (_("\
      --no-escape             do not use C escapes in output (default)\n"));
      printf (_("\
      --escape                use C escapes in output, no extended chars\n"));
      printf (_("\
      --force-po              write PO file even if empty\n"));
      printf (_("\
      --indent                indented output style\n"));
      printf (_("\
      --no-location           suppress '#: filename:line' lines\n"));
      printf (_("\
      --add-location          preserve '#: filename:line' lines (default)\n"));
      printf (_("\
      --strict                strict Uniforum output style\n"));
      printf (_("\
  -p, --properties-output     write out a Java .properties file\n"));
      printf (_("\
      --stringtable-output    write out a NeXTstep/GNUstep .strings file\n"));
      printf (_("\
  -w, --width=NUMBER          set output page width\n"));
      printf (_("\
      --no-wrap               do not break long message lines, longer than\n\
                              the output page width, into several lines\n"));
      printf (_("\
      --sort-output           generate sorted output\n"));
      printf (_("\
      --sort-by-file          sort output by file location\n"));
      printf ("\n");
      printf (_("\
Informative output:\n"));
      printf (_("\
  -h, --help                  display this help and exit\n"));
      printf (_("\
  -V, --version               output version information and exit\n"));
      printf ("\n");
      fputs (_("Report bugs to <bug-gnu-gettext@gnu.org>.\n"),
	     stdout);
    }

  exit (status);
}


/* Return 1 if FILENAME is contained in a list of filename patterns,
   0 otherwise.  */
static bool
filename_list_match (const string_list_ty *slp, const char *filename)
{
  size_t j;

  for (j = 0; j < slp->nitems; ++j)
    if (fnmatch (slp->item[j], filename, FNM_PATHNAME) == 0)
      return true;
  return false;
}


#ifdef EINTR

/* EINTR handling for close().
   These functions can return -1/EINTR even though we don't have any
   signal handlers set up, namely when we get interrupted via SIGSTOP.  */

static inline int
nonintr_close (int fd)
{
  int retval;

  do
    retval = close (fd);
  while (retval < 0 && errno == EINTR);

  return retval;
}
#define close nonintr_close

#endif


/* Process a string STR of size LEN bytes through grep, and return true
   if it matches.  */
static bool
is_string_selected (int grep_pass, const char *str, size_t len)
{
  const struct grep_task *gt = &grep_task[grep_pass];

  if (gt->pattern_count > 0)
    {
      size_t match_size;
      size_t match_offset;

      match_offset =
	gt->matcher->execute (gt->compiled_patterns, str, len,
			      &match_size, false);
      return (match_offset != (size_t) -1);
    }
  else
    return 0;
}


/* Return true if a message matches.  */
static bool
is_message_selected (const message_ty *mp)
{
  size_t i;
  const char *msgstr;
  size_t msgstr_len;
  const char *p;

  /* Always keep the header entry.  */
  if (mp->msgid[0] == '\0')
    return true;

  /* Test whether one of mp->filepos[] is selected.  */
  for (i = 0; i < mp->filepos_count; i++)
    if (filename_list_match (location_files, mp->filepos[i].file_name))
      return true;

  /* Test msgid and msgid_plural using the --msgid arguments.  */
  if (is_string_selected (0, mp->msgid, strlen (mp->msgid)))
    return true;
  if (mp->msgid_plural != NULL
      && is_string_selected (0, mp->msgid_plural, strlen (mp->msgid_plural)))
    return true;

  /* Test msgstr using the --msgstr arguments.  */
  msgstr = mp->msgstr;
  msgstr_len = mp->msgstr_len;
  /* Process each NUL delimited substring separately.  */
  for (p = msgstr; p < msgstr + msgstr_len; )
    {
      size_t length = strlen (p);

      if (is_string_selected (1, p, length))
	return true;

      p += length + 1;
    }

  /* Test translator comments using the --comment arguments.  */
  if (grep_task[2].pattern_count > 0
      && mp->comment != NULL && mp->comment->nitems > 0)
    {
      size_t length;
      char *total_comment;
      char *q;
      size_t j;
      bool selected;

      length = 0;
      for (j = 0; j < mp->comment->nitems; j++)
	length += strlen (mp->comment->item[j]) + 1;
      total_comment = (char *) xallocsa (length);

      q = total_comment;
      for (j = 0; j < mp->comment->nitems; j++)
	{
	  size_t l = strlen (mp->comment->item[j]);

	  memcpy (q, mp->comment->item[j], l);
	  q += l;
	  *q++ = '\n';
	}
      if (q != total_comment + length)
	abort ();

      selected = is_string_selected (2, total_comment, length);

      freesa (total_comment);

      if (selected)
	return true;
    }

  return false;
}


static void
process_message_list (const char *domain, message_list_ty *mlp)
{
  if (string_list_member (domain_names, domain))
    /* Keep all the messages in the list.  */
    ;
  else
    /* Keep only the selected messages.  */
    message_list_remove_if_not (mlp, is_message_selected);
}


static msgdomain_list_ty *
process_msgdomain_list (msgdomain_list_ty *mdlp)
{
  size_t k;

  for (k = 0; k < mdlp->nitems; k++)
    process_message_list (mdlp->item[k]->domain, mdlp->item[k]->messages);

  return mdlp;
}
